snyk-sast: added stats for Snyk scans

jperezdealgaba opened the pull request:
Solves: https://issues.redhat.com/browse/OSH-769

Adding the stats to snyk scans in the logs traces and as a result of the task.

# Before you complete this pull request ...

Look for any open pull requests in the repository with the title "e2e-tests update" and 
see if there are recent e2e-tests updates that will be applicable to your change.

jperezdealgaba commented:
@kdudka Would you mind checking this draft MR? This PR adds stats to the Snyk stats using the same calculations as we do in https://github.com/csutils/csmock/blob/main/py/common/snyk.py#L21

The stats are shown in the task logs and are also shown as a result of the task:
![Screenshot 2024-11-20 at 12 39 33](https://github.com/user-attachments/assets/18d1ea91-05c5-46c7-a7e3-0182845f6aa8)

Creating a new version for the task really difficults reviewing this code MR
The added code is from https://github.com/jperezdealgaba/build-definitions/pull/3/files#diff-9adcd5c3ca1989c09216495db837be670f48cc2e44adcef8821603fbeb54defdR203 to https://github.com/jperezdealgaba/build-definitions/pull/3/files#diff-9adcd5c3ca1989c09216495db837be670f48cc2e44adcef8821603fbeb54defdR2330 

Open questions:
- Should we create a new version of the task for this?
- I don't feel comfortable creating a single metadata file for each task, as we will need to create a lot of files adding overhead. Once many tasks show the stats, we can create a task the create the file with the results of all the SAST files.
- Review the code please


Example pipeline: https://konflux.apps.stone-prod-p02.hjvn.p1.openshiftapps.com/application-pipeline/workspaces/jperezde/applications/test-coverity/pipelineruns/osh-cli-container-konflux-test-2-on-pull-request-kbxf7
Thank you!

kdudka commented:
> Should we create a new version of the task for this?

I do not think that a new version is needed.  The change should be backward compatible.

> I don't feel comfortable creating a single metadata file for each task, as we will need to create a lot of files adding overhead. Once many tasks show the stats, we can create a task the create the file with the results of all the SAST files.

My idea was to embed the statistics directly into the SARIF file produced by each task.  This should be easy to implement with `csgrep --set-scan-prop=...`.  The code to obtain the numbers will be different for each scanner though.

jperezdealgaba commented:
@kdudka I did the changes. I didn't remove the JSON in the logs and in the results as I think they are useful to have the stats in the UI without downloading the results. If you think they don't just let me know and I will remove them.
The stats are now integrated in the file with the same naming as we had in OSH:
![Screenshot 2024-11-20 at 16 43 19](https://github.com/user-attachments/assets/c11d3bf7-fb15-4c7f-93b3-a78fcffa54b8)

I can not upload the SARIF file here, but this is the digest of the image: `quay.io/redhat-user-workloads/jperezde-tenant/test-coverity/osh-cli-container-konflux-test-2@sha256:b0f7ac7fc2b3ed8bbbf846e205cedae690fcf25a953759a29390208cbd945943`

Example pipeline: https://konflux.apps.stone-prod-p02.hjvn.p1.openshiftapps.com/application-pipeline/workspaces/jperezde/applications/test-coverity/pipelineruns/osh-cli-container-konflux-test-2-on-pull-request-nf2hp

Let me know what you think!

P.S: In future versions I won't create a new version but use the `0.3`

kdudka commented on task/sast-snyk-check/0.4/sast-snyk-check.yaml:
```diff
@@ -0,0 +1,280 @@
+              coverage_ratio=$((supported_files * 100 / total_files))
+          fi
+        
+          csgrep --mode=sarif --set-scan-prop snyk-scanned-files-coverage:${coverage_ratio} \
```
I would not run `csgrep` twice.  The additional decoding and encoding of SARIF could be expensive on scans that produce too many findings. We can simply append the arguments to the original invocation of `csgrep`.

jperezdealgaba replied:
I have modified the calls to `csgrep` and only one is done.

kdudka commented on task/sast-snyk-check/0.4/sast-snyk-check.yaml:
```diff
@@ -0,0 +1,280 @@
+    - description: Tekton task test output.
+      name: TEST_OUTPUT
+    - description: JSON value with the stats of the scan
+      name: SCAN_STATS
```
I would not introduce a separate result for this.  The information can be obtained from the resulting SARIF file (and I doubt that users would proactively use it for anything).  Let's keep the high-level interface simple.

jperezdealgaba replied:
Removed `SCAN_STATS` result and the interface

kdudka commented on task/sast-snyk-check/0.4/sast-snyk-check.yaml:
```diff
@@ -0,0 +1,280 @@
+          total_files=$(jq '[.runs[0].properties.coverage[].files] | add' "${SOURCE_CODE_DIR}"/sast_snyk_check_out.json)
+          supported_files=$(jq '[.runs[0].properties.coverage[] | select(.type == "SUPPORTED") | .files] | add' "${SOURCE_CODE_DIR}"/sast_snyk_check_out.json)
+
+          # We make sure the values are not null (jq returns null for empty lists)
```
What do you mean by `null` here?  An empty string?

jperezdealgaba replied:
Yes. If no supported/total files are found, jq returns an empty list. We convert it to 0.

I have modified the comment to make it clearer

kdudka reviewed (changes requested):
As said in my previous comment, I would not do the version bump for a backward-compatible change.  This would be unnecessarily disruptive for the users.

jperezdealgaba committed:
snyk-sast: added stats for Snyk scans

Solves: https://issues.redhat.com/browse/OSH-769

Adding the stats to snyk scans in the logs traces and as a result of the task.

kdudka reviewed (approved):
Looks good although I have not tested it myself.

Please do not forget to run `hack/generate-ta-tasks.sh` to update the `-oci-ta` copy of the task.

jperezdealgaba commented:
@kdudka Sure! This was mainly to discuss design things with you before creating the MR in upstream. I will do it as soon as possible.
I will also create the Coverity stats MR to be discussed with you and maybe it will be merged with the original Coverity MR...
